### PR TITLE
fix(swaps): prevent SwapFailurePacket crash

### PR DESCRIPTION
**Pull request by ghost**
In this commit we're fixing a bug where on 2 occasions it was possible
to crash the xud process because of type mismatch when serializing the
packet.

Furthermore, we're changing the call signature of `sendErrorToPeer.sendErrorToPeer` to an object in order to prevent such mistakes in the future.

Related issues: 
https://github.com/ExchangeUnion/xud/issues/984
https://github.com/ExchangeUnion/xud/issues/986

### CODE REVIEW

File: lib/swaps/Swaps.ts [hunk 1]
```diff
@@ -174,7 +174,10 @@ class Swaps extends EventEmitter {
   /**
    * Sends an error to peer. Sets reqId if packet is a response to a request.
    */
-  private sendErrorToPeer = async (peer: Peer, rHash: string, failureReason: SwapFailureReason, errorMessage?: string, reqId?: string) => {
+  private sendErrorToPeer = async ({
+    peer, rHash, failureReason = SwapFailureReason.UnknownError, errorMessage, reqId }:
+    {peer: Peer, rHash: string, failureReason?: SwapFailureReason, errorMessage?: string, reqId?: string,
```
Comment (sangaman) [inline lib/swaps/Swaps.ts]: These lines are a little bit hard to read, could you put the open and close `{}` on the same line? It looks off to me when one of the two braces is in-line with the properties but the other is not.

File: lib/swaps/Swaps.ts [hunk 2]
```diff
@@ -403,7 +406,12 @@ class Swaps extends EventEmitter {
 
     const rHash = requestPacket.body!.rHash;
     if (this.usedHashes.has(rHash)) {
-      await this.sendErrorToPeer(peer, requestPacket.body!.rHash, SwapFailureReason.PaymentHashReuse, undefined, requestPacket.header.id);
+      await this.sendErrorToPeer({
+        peer,
+        rHash: requestPacket.body!.rHash,
```
Comment (sangaman) [inline lib/swaps/Swaps.ts]: You can just write this as `rHash,` right? Since we declare it above on line 407.

File: lib/swaps/Swaps.ts [hunk 3]
```diff
@@ -448,21 +462,39 @@ class Swaps extends EventEmitter {
     // Make sure we are connected to lnd for both currencies
     if (!this.isPairSupported(deal.pairId)) {
       this.failDeal(deal, SwapFailureReason.SwapClientNotSetup);
-      await this.sendErrorToPeer(peer, deal.rHash, deal.failureReason!, deal.errorMessage, requestPacket.header.id);
+      await this.sendErrorToPeer({
+        peer,
+        rHash: deal.rHash,
```
Comment (sangaman) [inline lib/swaps/Swaps.ts]: I think this can also just be `rHash` as well.

File: lib/swaps/Swaps.ts [hunk 4]
```diff
@@ -448,21 +462,39 @@ class Swaps extends EventEmitter {
...
     }
 
     try {
       deal.makerToTakerRoutes = await takerSwapClient.getRoutes(takerAmount, takerPubKey);
     } catch (err) {
       this.failDeal(deal, SwapFailureReason.UnexpectedClientError, err.message);
-      await this.sendErrorToPeer(peer, deal.rHash, deal.failureReason!, deal.errorMessage, requestPacket.header.id);
+      await this.sendErrorToPeer({
+        peer,
+        rHash: deal.rHash,
```
Comment (sangaman) [inline lib/swaps/Swaps.ts]: Same here for `rHash`, and lines 493 and 508.

File: lib/swaps/Swaps.ts [hunk 5]
```diff
@@ -565,7 +603,11 @@ class Swaps extends EventEmitter {
       await takerSwapClient.addInvoice(deal.rHash, deal.takerAmount);
     } catch (err) {
       this.failDeal(deal, SwapFailureReason.UnexpectedClientError, err.message);
-      await this.sendErrorToPeer(peer, rHash, err.message);
+      await this.sendErrorToPeer({
+        peer,
+        rHash,
+        errorMessage: err.message,
```
Comment (sangaman) [inline lib/swaps/Swaps.ts]: I think we can set the failure reason to `SwapFailureReason.UnexpectedClientError` here.

File: lib/swaps/Swaps.ts [hunk 6]
```diff
@@ -582,7 +624,11 @@ class Swaps extends EventEmitter {
       await peer.sendPacket(new packets.SwapCompletePacket(responseBody));
     } catch (err) {
       this.failDeal(deal, SwapFailureReason.SendPaymentFailure, err.message);
-      await this.sendErrorToPeer(peer, rHash, err.message);
+      await this.sendErrorToPeer({
+        peer,
+        rHash,
+        errorMessage: err.message,
```
Comment (sangaman) [inline lib/swaps/Swaps.ts]: Similarly we can set `SwapFailureReason.SendPaymentFailure` for the reason here.

**Review by sangaman — COMMENTED** [review]
Thanks for diagnosing this. I just have some comments/suggestions that are mainly stylistic.

**Comment by erkarl** [comment]
@sangaman thanks for the feedback. Changes made.

**Review by sangaman — APPROVED** [review]
Awesome, thank you!

**Pull request merged**